dali/operators/segmentation/select_masks.cc

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/segmentation/select_masks.h"
+
+namespace dali {
+
+DALI_SCHEMA(segmentation__SelectMasks)
+    .DocStr(R"(Selects a subset of mask by their mask ids.
+
+The operator expects ``mask_meta`` and ``mask_coords``, containing segmentation mask.
+Each sample can contain several mask, and each mask can be composed by several polygons.
+The format of the data is the same as the one produced by COCOReader:
+
+- ``mask_meta`` -> 2D data where each row describes a single mask with three values, ``mask_idx, start_idx, end_idx``,
+where ``start_idx`` and ``end_idx`` refer to indices in ``mask_coords``.
+
+- ``mask_coords``-> 2D data where each row contains a single ``x, y`` coordinate.
+
+The operator receives a list of mask ids that are to be selected for each sample, and produces
+output ``mask_meta`` and ``mask_coords`` where only the given mask ids are preserved.
+
+**Examples:**
+
+Let us assume the following input mask, where symbolic coordinates are used for a clearer example::
+
+    ``mask_coords`` = [[x0, y0], [x1, y1], [x2, y2], [x3, y3], [x4, y4], [x5, y5], [x6, y6], [x7, y7], [x8, y8], [x9, y9]]
+    ``mask_meta`` = [[0, 0, 2], [1, 3, 6], [2, 7, 9]
+
+Example 1: Selecting a single mask with id ``1``, maintaining the original id::
+
+    ``mask_ids`` = [1], ``reindex_masks`` = False
+    ``out_mask_meta`` = [[1, 0, 4]]
+    ``out_mask_coords`` = [[x3, y3], [x4, y4], [x5, y5], [x6, y6]]
+
+Example 2: Selecting two out of the three mask, reindexing the mask ids to follow the order of the input mask ids::
+
+``mask_ids`` = [2, 0]
```

Comment:
Do we want to allow such inversion, what is the use case?

Reply:
The thing is that the mask_ids are actually indices of the bounding boxes in the batch.

Let's say we had three bounding boxes [A, B, C] with corresponding masks ids [0, 1, 2].
Now random bbox crop filtered out B and we are left with [A, B] and an output describing the indices of the bounding boxes that stayed: [0, 2].

What we want now is select the masks that refer to those bounding boxes that stayed, so we filter mask_ids [0, 2]. However, the bounding boxes now are two, so we have indices [0, 1]. To be able to map a mask to a bounding box the mask id should have the proper index, so we do this mapping: mask_id remains 0, but mask_id is now 1 because that's the index of the bounding box in the cropped image.

Reply:
I understand the reason for reindexing, but I was asking about the [0, 2] case. Do we expect operators to change the order of bboxes?

Reply:
I think the order doesn't matter much. I mean, there is no special reason to reorder them, it is just that we are processing them as they appear in the mask_ids input.